### PR TITLE
[tmva][sofie]  Add new operators and several improvements in SOFIE

### PR DESCRIPTION
**Pull request by lmoneta**
# This Pull request adds (see commits logs for more details)

- TopK, Tile and Split  operators
- If operator and support for subgraphs by parsing them from ONNX and dealing with subgraphs in RModel
- Add MatMul support for dim=1 tensors 
- Support output of operators as constant tensors

### CODE REVIEW

File: tmva/sofie/inc/TMVA/ROperator_TopK.hxx [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+      // fShapeX = model.GetTensorShape(fNX); //  [ m x n x o x p ... ]
+      // if(k[0]>=fShapeX.size()){
+      //    throw
+      //       std::runtime_error("TMVA::SOFIE ONNX TopK op k = "+ std::to_string(k[0]) +"value exeeds size of tensor " +fNX+" of size "+fShapeX.size()+" .");
+      // }
+      // fShapeY.push_back(2);
+      // for (auto i : fShapeX)
+      //    fShapeY.push_back(i); //  [ 2 x m x n x o x p ... ]
+      // size_t axis = fAttrAxis < 0 ? fShapeX.size() + fAttrAxis : fAttrAxis;
+      // fShapeY[axis] = k[0]; //  [ 2 x m x n x K x p ... ]
```
Comment (sanjibansg) [inline tmva/sofie/inc/TMVA/ROperator_TopK.hxx:96]: same suggestion with the commented code here

File: tmva/sofie/inc/TMVA/ROperator_TopK.hxx [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+      fShapeY=ShapeInference({fShapeX,fShapeK})[0];
+
+      // for(int i=0;i<fShapeX.size();i++)
+      // std::cout<<fShapeX[i]<<" ";
+      // std::cout<<"\ny size -> "<<fShapeY.size()<<std::endl;
```
Comment (sanjibansg) [inline tmva/sofie/inc/TMVA/ROperator_TopK.hxx:101]:
```suggestion

```

File: tmva/sofie_parsers/src/ParseConcat.cxx [hunk 3]
```diff
@@ -37,12 +37,14 @@ ParserFuncSignature ParseConcat = [](RModelParser_ONNX &parser, const onnx::Node
...
          attr_new_axis = nodeproto.attribute(i).i();
    }
-   switch (input_type) {
-   case ETensorType::FLOAT: op.reset(new ROperator_Concat<float>(inputs, attr_axis, attr_new_axis, output_name)); break;
-   default:
-      throw std::runtime_error("TMVA::SOFIE - Unsupported - Operator Concat does not yet support input type " +
-                               std::to_string(static_cast<int>(input_type)));
-   }
+   //switch (input_type) {
+   //case ETensorType::FLOAT:
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/ParseConcat.cxx:41]:
```suggestion

```

File: tmva/sofie_parsers/src/ParseConcat.cxx [hunk 4]
```diff
@@ -37,12 +37,14 @@ ParserFuncSignature ParseConcat = [](RModelParser_ONNX &parser, const onnx::Node
...
+   op.reset(new ROperator_Concat(inputs, attr_axis, attr_new_axis, output_name));
+   //break;
+   //default:
+   //   throw std::runtime_error("TMVA::SOFIE - Unsupported - Operator Concat does not yet support input type " +
+   //                            std::to_string(static_cast<int>(input_type)));
+  // }
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/ParseConcat.cxx:47]:
```suggestion

```

File: tmva/sofie_parsers/src/ParseReshape.cxx [hunk 5]
```diff
@@ -47,17 +47,14 @@ ParserFuncSignature ParseReshape = [](RModelParser_ONNX &parser, const onnx::Nod
...
-                               std::to_string(static_cast<int>(input_type)));
-   }
+
+   if (attr_axes.empty())
+      op.reset(new ROperator_Reshape(opMode, attr_value, input_name, shape_name, output_name));
+   else // for old Squeeze and Unsqueeze
+      op.reset(new ROperator_Reshape(opMode, attr_axes, input_name, output_name));
+
+   //   throw std::runtime_error("TMVA::SOFIE - Unsupported - Operator Reshape does not yet support input type " +
+   //                            ConvertTypeToString(input_type));
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/ParseReshape.cxx:57]:
```suggestion

```

File: tmva/sofie_parsers/src/ParseSlice.cxx [hunk 6]
```diff
@@ -48,28 +48,28 @@ ParserFuncSignature ParseSlice = [](RModelParser_ONNX &parser, const onnx::NodeP
 
    std::unique_ptr<ROperator> op;
    std::string output_name = nodeproto.output(0);
-   switch (input_type) {
-   case ETensorType::FLOAT:
+   //switch (input_type) {
+   //case ETensorType::FLOAT:
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/ParseSlice.cxx:52]:
```suggestion
```

File: tmva/sofie_parsers/src/ParseSlice.cxx [hunk 7]
```diff
@@ -48,28 +48,28 @@ ParserFuncSignature ParseSlice = [](RModelParser_ONNX &parser, const onnx::NodeP
...
-      break;
-   default:
-      throw std::runtime_error("TMVA::SOFIE - Unsupported - Operator Slice does not yet support input type " +
-                               std::to_string(static_cast<int>(input_type)));
-   }
+      //break;
+   //default:
+   //   throw std::runtime_error("TMVA::SOFIE - Unsupported - Operator Slice does not yet support input type " +
+   //                            std::to_string(static_cast<int>(input_type)));
+   //}
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/ParseSlice.cxx:72]:
```suggestion
 
```

File: tmva/sofie_parsers/src/ParseTopK.cxx [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+   if (!parser.IsRegisteredTensorType(outputInd_name)) {
+      parser.RegisterTensorType(outputInd_name, ETensorType::INT64);
+   }
+
+   return op;
+};
+
+} // namespace SOFIE
+} // namespace Experimental
+} // namespace TMVA
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/ParseTopK.cxx:56]: we might need to add a newline here?

File: tmva/sofie_parsers/src/ParseTile.cxx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+   if (!parser.IsRegisteredTensorType(output_name)) {
+      parser.RegisterTensorType(output_name, input_type);
+   }
+
+   return op;
+};
+
+} // namespace SOFIE
+} // namespace Experimental
+} // namespace TMVA
```
Comment (sanjibansg) [inline tmva/sofie_parsers/src/ParseTile.cxx:47]: same suggestion with the newline to follow the overall code style

**Review by sanjibansg — CHANGES_REQUESTED** [review]
LGTM! 
Thanks for the implementation, all this is very helpful!
Just some small suggestions on commented code and codestyle

**Review by sanjibansg — APPROVED** [review]
LGTM!
Thanks for this!

**Review by dpiparo — APPROVED** [review]
lgtm!

**Pull request merged**